### PR TITLE
Simulated Annealing

### PR DESCRIPTION
**Pull request by AmeenUrRehman**
Added the annealing_box curriculum
updated the test file with some customized input
update the curricula init

### CODE REVIEW

File: syllabus/curricula/annealing_box.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+class AnnealingBoxCurriculum(Curriculum):
+    REQUIRES_STEP_UPDATES = True
+    REQUIRES_CENTRAL_UPDATES = False
+
+    def __init__(
+        self,
+        *curriculum_args,
+        start_values: List[SupportsFloat],
+        end_values: List[SupportsFloat],
+        steps: int = 5,
```
Comment (RyanNavillus) [inline syllabus/curricula/annealing_box.py:17]: I would call this `total_steps` and remove the default value, since it will be very different depending on the environment, algorithm, etc.

Also, could we make this a list? That way we can anneal one value for the first half of training, another for the entirety of training, etc.

File: tests/multiprocessing_smoke_tests.py [hunk 2]
```diff
@@ -19,6 +19,11 @@
         (LearningProgressCurriculum, create_nethack_env, (nethack_env.task_space,), {}),
         (PrioritizedLevelReplay, create_nethack_env, (nethack_env.task_space,), {"device": "cpu", "suppress_usage_warnings": True, "num_processes": N_ENVS}),
         (SimpleBoxCurriculum, create_cartpole_env, (cartpole_env.task_space,), {}),
+        (AnnealingBoxCurriculum, create_cartpole_env, (cartpole_env.task_space,), {
+            'start_values': [0.2, 0.5],
+            'end_values': [0.8, 0.9],
+            'steps': 10
+        }),
```
Comment (RyanNavillus) [inline tests/multiprocessing_smoke_tests.py:26]: Have you tried running this? Does it work? Also please try to add some more thorough unit tests in a separate file (check that the annealed values are what you would expect after a specific number of steps).

File: syllabus/curricula/annealing_box.py (hunk 1 above)
Comment (AmeenUrRehman) [reply syllabus/curricula/annealing_box.py]: yeh, I've updated this

File: tests/multiprocessing_smoke_tests.py (hunk 2 above)
Comment (AmeenUrRehman) [reply tests/multiprocessing_smoke_tests.py:26]: Hi, should I have to create a custom test using unittest with a different file??

Comment (RyanNavillus) [reply tests/multiprocessing_smoke_tests.py:26]: Yes, it would be great if you could create a specific test file for simulated annealing at write some common sense tests to make sure the internals work as expected.

File: syllabus/curricula/annealing_box.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        super().__init__(*curriculum_args, **curriculum_kwargs)
+        assert isinstance(
+            self.task_space.gym_space, Box
+        ), "AnnealingBoxCurriculum only supports Box task spaces."
+
+        self.start_values = np.array(start_values, dtype=np.float32)
+        self.end_values = np.array(end_values, dtype=np.float32)
+        self.total_steps = np.array(total_steps, dtype=int)
+
+        assert len(self.start_values) == len(self.end_values), "Length of start_values and end_values must be the same."
```
Comment (RyanNavillus) [inline syllabus/curricula/annealing_box.py:29]: Maybe include an assert to check that all elements of total_steps > 0

**Review by RyanNavillus — COMMENTED** [review]
Thanks for making those changes!

File: syllabus/curricula/annealing_box.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    def sample(self, k: int = 1) -> Union[List, Any]:
+        """
+        Sample k tasks from the curriculum.
+        """
+        # Linear annealing from start_values to end_values
+        annealed_values = (
+                self.start_values + (self.end_values - self.start_values) * (
+                    float(self.current_step) / float(self.total_steps))
+        )
```
Comment (RyanNavillus) [inline syllabus/curricula/annealing_box.py:48]:
```suggestion
        annealed_values = (
                self.start_values + (self.end_values - self.start_values)
                * (float(self.current_step) / float(self.total_steps))
        )
```
I think this is valid syntax and easier to read

File: tests/annealing_box_unittests.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+class TestAnnealingBoxCurriculum(unittest.TestCase):
+
+    def test_annealing_values(self):
+        cartpole_env = create_cartpole_env()
+        curriculum = AnnealingBoxCurriculum(
+            task_space=cartpole_env.task_space,
+            start_values=[0.2, 0.5],
+            end_values=[0.8, 0.9],
+            total_steps=10
```
Comment (RyanNavillus) [inline tests/annealing_box_unittests.py:14]: total_steps should be a list here as well

File: tests/annealing_box_unittests.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            [0.456, 0.56],
+            [0.584, 0.59],
+            [0.712, 0.62],
+            [0.84, 0.65],
+            [0.968, 0.68],
+            [1.096, 0.71],
+            [1.224, 0.74],
+            [1.352, 0.77],
+            [1.48, 0.8]
+        ]
```
Comment (RyanNavillus) [inline tests/annealing_box_unittests.py:30]: These expected values seem very wrong? The final values should be 0.8, 0.9

So something like:
[0.20, 0.50]
[0.26, 0.54]
[0.32, 0.58]
[0.38, 0.62]
[0.44, 0.66]
[0.50, 0.70]
[0.56, 0.74]
[0.62, 0.78]
[0.68, 0.82]
[0.74, 0.86]
[0.80, 0.90]
[0.80, 0.90]
[0.80, 0.90]

File: tests/annealing_box_unittests.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import unittest
+import numpy as np
+from syllabus.tests import create_cartpole_env
+from syllabus.curricula import AnnealingBoxCurriculum
+
+class TestAnnealingBoxCurriculum(unittest.TestCase):
```
Comment (RyanNavillus) [inline tests/annealing_box_unittests.py:6]: Really nice test! I didn't even know that numpy.testing existed. Please also add test cases for negative start_values/end_values, and reverse annealing (e.g. 0.7 -> 0.1 and 0.7 -> -0.1).

File: syllabus/curricula/annealing_box.py (hunk 4 above)
Comment (RyanNavillus) [inline syllabus/curricula/annealing_box.py:48]: I forgot to mention, after total_steps is reached, the curriculum should just return the end_value. So something like:
```suggestion
        annealed_values = (
                self.start_values + (self.end_values - self.start_values) * (
                    float(np.min(self.current_step, self.total_steps)) / float(self.total_steps))
        )
```

(Not sure if this code actually works)

**Review by RyanNavillus — COMMENTED** [review]
Looks great overall! As I was reading over your code I remembered some thing I should have asked you to do earlier. I think this should be the final set of changes before we merge though.

**Comment by RyanNavillus** [comment]
Oh and please also add this curriculum to `test/multiprocessing_smoke_tests.py` All that does is check whether the curriculum runs without crashing.
